File: code/backend/src/main/java/com/taskflow/modules/workspace/service/WorkspaceService.java
package com.taskflow.modules.workspace.service;

import com.taskflow.modules.workspace.dto.CreateWorkspaceRequest;
import com.taskflow.modules.workspace.dto.UpdateWorkspaceRequest;
import com.taskflow.modules.workspace.dto.WorkspaceDto;
import com.taskflow.modules.workspace.dto.WorkspaceMemberDto;

import java.util.List;
import java.util.UUID;

/**
 * Domain Service interface for managing workspace creation, membership roles, settings, and soft-deletion.
 */
public interface WorkspaceService {

    /**
     * Creates a new workspace and registers the creator as OWNER.
     *
     * @param ownerId UUID identifier of the workspace owner
     * @param request creation request payload
     * @return WorkspaceDto of the created workspace
     */
    WorkspaceDto createWorkspace(UUID ownerId, CreateWorkspaceRequest request);

    /**
     * Retrieves all active workspaces accessible by the user.
     *
     * @param userId UUID identifier of the requesting user
     * @return list of accessible WorkspaceDto instances
     */
    List<WorkspaceDto> getUserWorkspaces(UUID userId);

    /**
     * Retrieves detailed information of a specific workspace, verifying membership access.
     *
     * @param userId      UUID identifier of the requesting user
     * @param workspaceId UUID identifier of the target workspace
     * @return WorkspaceDto instance
     */
    WorkspaceDto getWorkspaceDetails(UUID userId, UUID workspaceId);

    /**
     * Updates workspace settings (name, description, theme color).
     * Requires OWNER or ADMIN role.
     *
     * @param userId      UUID identifier of the requesting user
     * @param workspaceId UUID identifier of the target workspace
     * @param request     update payload
     * @return updated WorkspaceDto instance
     */
    WorkspaceDto updateWorkspace(UUID userId, UUID workspaceId, UpdateWorkspaceRequest request);

    /**
     * Soft-deletes a workspace. Requires OWNER role.
     *
     * @param userId      UUID identifier of the requesting user
     * @param workspaceId UUID identifier of the target workspace
     */
    void deleteWorkspace(UUID userId, UUID workspaceId);

    /**
     * Retrieves list of active workspace members and assigned roles.
     *
     * @param userId      UUID identifier of the requesting user
     * @param workspaceId UUID identifier of the target workspace
     * @return list of WorkspaceMemberDto instances
     */
    List<WorkspaceMemberDto> getWorkspaceMembers(UUID userId, UUID workspaceId);
}

File: code/frontend/src/features/workspace/services/workspace-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type { CreateWorkspaceInput, UpdateWorkspaceInput, WorkspaceDto, WorkspaceMemberDto } from '../types';

export const workspaceService = {
  createWorkspace: async (data: CreateWorkspaceInput): Promise<WorkspaceDto> => {
    const res = await apiClient.post<ApiResponse<WorkspaceDto>>('/workspaces', data);
    return res.data.data;
  },

  getUserWorkspaces: async (): Promise<WorkspaceDto[]> => {
    const res = await apiClient.get<ApiResponse<WorkspaceDto[]>>('/workspaces');
    return res.data.data;
  },

  getWorkspaceDetails: async (workspaceId: string): Promise<WorkspaceDto> => {
    const res = await apiClient.get<ApiResponse<WorkspaceDto>>(`/workspaces/${workspaceId}`);
    return res.data.data;
  },

  updateWorkspace: async (workspaceId: string, data: UpdateWorkspaceInput): Promise<WorkspaceDto> => {
    const res = await apiClient.put<ApiResponse<WorkspaceDto>>(`/workspaces/${workspaceId}`, data);
    return res.data.data;
  },

  deleteWorkspace: async (workspaceId: string): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`/workspaces/${workspaceId}`);
  },

  getWorkspaceMembers: async (workspaceId: string): Promise<WorkspaceMemberDto[]> => {
    const res = await apiClient.get<ApiResponse<WorkspaceMemberDto[]>>(`/workspaces/${workspaceId}/members`);
    return res.data.data;
  },
};
